File: native/RtlFu/LTageBp.scala
// L-TAGE branch predictor — the first RTL predictor exercising the speculative-history
// port contract, driven through RtlFfiHistoryBranchPredictor (src/Core/Mechanism/RtlFu)
// via rtl_hbp_shim.cpp.
//
// Deliberately mirrors Horologium's C# LTagePredictor bit-for-bit so the differential
// co-sim test can demand identical predictions under fetch/commit/flush/partial-squash
// sequences:
//   - 4096-entry bimodal base (2-bit, init weakly-not-taken — stored biased so the
//     zero-initialized memory reads as 1) + 4 tagged tables (512 × {9b tag, 3b ctr,
//     2b u, valid}; the C# ctr=3 construction init is dead state, allocation always
//     writes ctr explicitly, so zero-init valid=0 memories match)
//   - geometric history lengths 8/13/21/34 over a 34-bit GHR; folded indices/tags
//     recomputed combinationally from the GHR exactly like the C# FoldHist
//   - longest-match provider/alt selection, RRIP-style usefulness, allocate-on-
//     mispredict in the shortest longer-history table with u=0, else decay
//   - 32-entry loop predictor overlay with trip-count confidence
//   - speculative GHR contract: predict indexes the working GHR; update trains
//     against the committed shadow and advances it; spec_update folds a predicted
//     direction at fetch; recover restores the committed shadow; history/restore
//     expose the working GHR as the checkpoint (folds derive from it, so a single
//     34-bit value is a complete history snapshot — no checkpoint RAM needed)
//
// Direction-only: targets live in the C# wrapper's BTB dictionary (the C# predictor
// uses an unbounded Dictionary, which RTL cannot mirror).
//
// State is held in Mems (combinational read) so the generated Verilog stays compact.
//
// scala-cli project directives live in DivUnit.scala. Regenerate generated/LTageBp.sv
// with ./generate.sh after editing.

package rtlfu

import chisel3._
import chisel3.util._

class LTageBp extends Module {
  private val NumTables      = 4
  private val TableIndexBits = 9
  private val BaseIndexBits  = 12
  private val TagWidth       = 9
  private val MaxHist        = 34
  private val LoopIndexBits  = 5
  private val LoopTagWidth   = 10
  private val LoopConfidence = 4
  private val HistLengths    = Seq(8, 13, 21, 34)

  val io = IO(new Bundle {
    val predPc    = Input(UInt(64.W))
    val predTaken = Output(Bool())

    val updValid = Input(Bool())
    val updPc    = Input(UInt(64.W))
    val updTaken = Input(Bool())

    val specValid = Input(Bool())
    val specTaken = Input(Bool())

    val recoverValid = Input(Bool())

    val histOut   = Output(UInt(64.W))
    val restValid = Input(Bool())
    val restHist  = Input(UInt(64.W))
    val restTaken = Input(Bool())
  })

  class LoopEntry extends Bundle {
    val tag         = UInt(LoopTagWidth.W)
    val learnedIter = UInt(16.W)
    val currentIter = UInt(16.W)
    val confCount   = UInt(8.W)
    val confident   = Bool()
  }

  val ghr          = RegInit(0.U(MaxHist.W)) // working (speculative) history
  val committedGhr = RegInit(0.U(MaxHist.W)) // architectural shadow, advanced at update
  val specMode     = RegInit(false.B)        // latches once speculative updates arrive

  val baseMem = Mem(1 << BaseIndexBits, UInt(2.W)) // stores ctr ^ 1 → zero-init reads as 1
  val tagMems = Seq.fill(NumTables)(Mem(1 << TableIndexBits, UInt(TagWidth.W)))
  val ctrMems = Seq.fill(NumTables)(Mem(1 << TableIndexBits, UInt(3.W)))
  val uMems   = Seq.fill(NumTables)(Mem(1 << TableIndexBits, UInt(2.W)))
  val vMems   = Seq.fill(NumTables)(Mem(1 << TableIndexBits, Bool()))
  val loopMem = Mem(1 << LoopIndexBits, new LoopEntry)

  // ── Fold / index / tag helpers (mirror the C# FoldHist/TageIdx/TageTag) ─────

  private def fold(hist: UInt, histLen: Int, outBits: Int): UInt = {
    val masked = hist(histLen - 1, 0)
    (0 until histLen by outBits)
      .map { sh =>
        val hi = math.min(sh + outBits, histLen) - 1
        masked(hi, sh).pad(outBits)
      }
      .reduce(_ ^ _)
  }

  private def tageIdx(pc: UInt, hist: UInt, t: Int): UInt =
    ((pc >> 2)(TableIndexBits - 1, 0) ^ fold(hist, HistLengths(t), TableIndexBits))(TableIndexBits - 1, 0)

  private def tageTag(pc: UInt, hist: UInt, t: Int): UInt = {
    val f1 = fold(hist, HistLengths(t), TagWidth)
    val f2 = fold(hist, HistLengths(t) - 1, TagWidth - 1)
    ((pc >> 2)(TagWidth - 1, 0) ^ f1 ^ (f2 << 1).pad(TagWidth)(TagWidth - 1, 0))(TagWidth - 1, 0)
  }

  private def baseIdx(pc: UInt): UInt = (pc >> 2)(BaseIndexBits - 1, 0)
  private def baseCtr(pc: UInt): UInt = baseMem(baseIdx(pc)) ^ 1.U(2.W)

  private def loopIdx(pc: UInt): UInt = (pc >> 2)(LoopIndexBits - 1, 0)

  private def loopTag(pc: UInt): UInt = {
    val x = pc >> 2
    (x ^ (x >> LoopIndexBits))(LoopTagWidth - 1, 0)
  }

  // Scan tables shortest→longest: provider = longest match, alt = previous best.
  // provNum encodes provider+1 (0 = base). Mirrors the C# TageLookup fold.
  private def tageLookup(pc: UInt, hist: UInt): (UInt, Bool, Bool, Vec[UInt]) = {
    val idx  = VecInit((0 until NumTables).map(t => tageIdx(pc, hist, t)))
    val base = baseCtr(pc) >= 2.U
    var prov: UInt  = 0.U(3.W)
    var provP: Bool = base
    var altP: Bool  = base
    for (t <- 0 until NumTables) {
      val hit  = vMems(t)(idx(t)) && tagMems(t)(idx(t)) === tageTag(pc, hist, t)
      val pred = ctrMems(t)(idx(t)) >= 4.U
      altP = Mux(hit, provP, altP)
      provP = Mux(hit, pred, provP)
      prov = Mux(hit, (t + 1).U(3.W), prov)
    }
    (prov, provP, altP, idx)
  }

  // ── Predict (combinational, no state change) ────────────────────────────────

  private val (_, predProvPred, _, _) = tageLookup(io.predPc, ghr)
  private val ple                     = loopMem(loopIdx(io.predPc))
  private val loopHit                 = ple.tag === loopTag(io.predPc) && ple.confident
  io.predTaken := Mux(loopHit, ple.currentIter < ple.learnedIter, predProvPred)

  io.histOut := ghr

  // ── Update (trains against the committed shadow; one clock edge) ────────────

  private val (uProv, uProvPred, uAltPred, uIdx) = tageLookup(io.updPc, committedGhr)
  private val mispredict                         = uProvPred =/= io.updTaken

  // Allocation: shortest longer-history table with u == 0; else decay u in that range.
  private val eligible = VecInit(
    (0 until NumTables).map(t => (t + 1).U > uProv && uMems(t)(uIdx(t)) === 0.U)
  )
  private val anyAlloc = eligible.asUInt.orR
  private val allocOH  = PriorityEncoderOH(eligible.asUInt)

  when(io.updValid) {
    for (t <- 0 until NumTables) {
      val isProvider = uProv === (t + 1).U
      val ctr        = ctrMems(t)(uIdx(t))
      val u          = uMems(t)(uIdx(t))
      when(isProvider) {
        ctrMems(t).write(
          uIdx(t), Mux(io.updTaken, Mux(ctr === 7.U, 7.U, ctr + 1.U), Mux(ctr === 0.U, 0.U, ctr - 1.U))
        )
        when(uProvPred =/= uAltPred) {
          uMems(t).write(
            uIdx(t),
            Mux(uProvPred === io.updTaken, Mux(u === 3.U, 3.U, u + 1.U), Mux(u === 0.U, 0.U, u - 1.U))
          )
        }
      }.elsewhen(mispredict && anyAlloc && allocOH(t)) {
        tagMems(t).write(uIdx(t), tageTag(io.updPc, committedGhr, t))
        ctrMems(t).write(uIdx(t), Mux(io.updTaken, 4.U, 3.U)) // weakly in correct direction
        uMems(t).write(uIdx(t), 0.U)
        vMems(t).write(uIdx(t), true.B)
      }.elsewhen(mispredict && !anyAlloc && (t + 1).U > uProv && u > 0.U) {
        uMems(t).write(uIdx(t), u - 1.U)
      }
    }

    when(uProv === 0.U) {
      val b = baseCtr(io.updPc)
      val n = Mux(io.updTaken, Mux(b === 3.U, 3.U, b + 1.U), Mux(b === 0.U, 0.U, b - 1.U))
      baseMem.write(baseIdx(io.updPc), n ^ 1.U(2.W))
    }

    // ── Loop predictor (mirrors the C# UpdateLoop cases exactly) ──────────────
    val li       = loopIdx(io.updPc)
    val lt       = loopTag(io.updPc)
    val le       = loopMem(li)
    val tagMatch = le.tag === lt

    val fresh = Wire(new LoopEntry) // zeroed replacement entry with the new tag
    fresh.tag         := lt
    fresh.learnedIter := 0.U
    fresh.currentIter := 0.U
    fresh.confCount   := 0.U
    fresh.confident   := false.B

    val next = WireDefault(le)
    when(tagMatch) {
      when(io.updTaken) {
        next.currentIter := le.currentIter + 1.U
      }.otherwise {
        when(!le.confident) {
          when(le.learnedIter === 0.U || le.learnedIter === le.currentIter) {
            next.learnedIter := le.currentIter
            val cc = Mux(le.confCount === 255.U, le.confCount, le.confCount + 1.U)
            next.confCount := cc
            next.confident := cc >= LoopConfidence.U
          }.otherwise {
            next := fresh // inconsistent trip count — restart tracking at this slot
          }
        }.elsewhen(le.currentIter =/= le.learnedIter) {
          next.confident   := false.B
          next.confCount   := 0.U
          next.learnedIter := le.currentIter
        }
        next.currentIter := 0.U
      }
      loopMem.write(li, next)
    }.elsewhen(io.updTaken) {
      val start = WireDefault(fresh) // first taken sighting at this slot: start tracking
      start.currentIter := 1.U
      loopMem.write(li, start)
    }

    // ── History ───────────────────────────────────────────────────────────────
    val newCommitted = ((committedGhr << 1) | io.updTaken.asUInt)(MaxHist - 1, 0)
    committedGhr := newCommitted
    ghr          := Mux(specMode, ghr, newCommitted)
  }

  when(io.specValid) {
    ghr      := ((ghr << 1) | io.specTaken.asUInt)(MaxHist - 1, 0)
    specMode := true.B
  }

  when(io.recoverValid) {
    ghr := committedGhr
  }

  when(io.restValid) {
    ghr      := ((io.restHist << 1) | io.restTaken.asUInt)(MaxHist - 1, 0)
    specMode := true.B
  }
}

/** Emits generated/LTageBp.sv. Run via generate.sh. */
object GenerateLTage extends App {
  _root_.circt.stage.ChiselStage.emitSystemVerilogFile(
    new LTageBp,
    Array("--target-dir", "generated"),
    Array("--lowering-options=disallowLocalVariables,disallowPackedArrays", "--strip-debug-info")
  )
}

File: src/Core/Mechanism/BranchPred/LTageBp.cs
namespace Mechanism.BranchPred;

/// <summary>
///     L-TAGE: TAGE branch predictor with a loop predictor overlay.
///     <para>
///         TAGE uses a bimodal base and N tagged tables with geometrically increasing
///         history lengths, each entry holding a partial tag, a 3-bit saturating
///         prediction counter, and a 2-bit usefulness counter. Longest-matching-history
///         wins. On misprediction an entry is allocated in the shortest longer-history
///         table with u=0; if none is available, usefulness bits are decayed.
///     </para>
///     <para>
///         The loop predictor tracks branches with a stable trip count. Once a branch
///         has exited a loop the same number of times on LoopConfThreshold consecutive
///         invocations, it takes over from TAGE and predicts taken/not-taken exactly.
///     </para>
/// </summary>
public class LTageBp : IBranchPredictor {
    // ── TAGE parameters ───────────────────────────────────────────────────────
    /// <summary>Number of tagged TAGE history tables (excluding the bimodal base).</summary>
    protected const int NumTables = 4;

    private const int TableIndexBits = 9; // 512 entries per tagged table
    private const int BaseIndexBits = 12; // 4096-entry bimodal base

    /// <summary>Bit width of the partial tag stored in each TAGE entry.</summary>
    private const int TagWidth = 9; // bits of partial tag per entry

    /// <summary>Width, in bits, of the global history register.</summary>
    protected const int MaxHist = 34;

    // ── Loop predictor parameters ─────────────────────────────────────────────
    private const int LoopIndexBits = 5; // 32-entry loop table
    private const int LoopTagWidth = 10;
    private const int LoopConfidence = 4; // consistent exits before confident

    // Geometrically increasing history lengths (~1.6×)
    private static readonly int[] HistLengths = [8, 13, 21, 34,];

    // ── Storage ───────────────────────────────────────────────────────────────
    private readonly byte[] _base; // 2-bit counters (taken ≥ 2)

    private readonly Dictionary<ulong, ulong> _btb = new();
    private readonly LoopEntry[] _loop;   // [1 << LoopIndexBits]
    private readonly TageEntry[][] _tage; // [NumTables][1 << TableIndexBits]

    // Architectural history shadow: advanced only when a branch retires (in Update, with the
    // true outcome). Because commit and fetch are in-order, a branch that commits had every
    // older branch on its path predicted correctly, so this equals that branch's predict-time
    // history. Update swaps it into Ghr to train tables against the history the branch saw.
    private ulong _committedGhr;

    // Latches true once the pipeline drives speculative history (out-of-order). Until then
    // Update keeps Ghr == _committedGhr, so the commit-time-history behavior is bit-identical.
    private bool _speculative;

    /// <summary>
    ///     Working global history register used for indexing at Predict time. In an
    ///     out-of-order pipeline this is the *speculative* history: advanced at fetch by
    ///     <see cref="SpeculativeHistoryUpdate" /> with predicted directions, and restored from
    ///     <see cref="_committedGhr" /> on a flush. When no speculative updates arrive (in-order
    ///     pipelines), it stays in lock-step with <see cref="_committedGhr" />.
    /// </summary>
    protected ulong Ghr; // global history, LSB = most recent

    /// <summary>
    ///     Constructs an L-TAGE predictor.
    /// </summary>
    public LTageBp() {
        _base = new byte[1 << LTageBp.BaseIndexBits];
        Array.Fill(_base, (byte)1); // weakly not-taken

        _tage = new TageEntry[LTageBp.NumTables][];
        for (var t = 0; t < LTageBp.NumTables; t++) {
            _tage[t] = new TageEntry[1 << LTageBp.TableIndexBits];
            for (var i = 0; i < _tage[t].Length; i++) _tage[t][i].Ctr = 3; // weakly not-taken (3-bit: taken ≥ 4)
        }

        _loop = new LoopEntry[1 << LTageBp.LoopIndexBits];
    }

    // ── IBranchPredictor ──────────────────────────────────────────────────────

    /// <inheritdoc />
    public BranchPrediction Predict(ulong pc, (ulong Value, bool HasValue) knownTarget = default) {
        TageLookup(pc, out int provider, out bool tagePred, out _);
        bool pred = ResolvePrediction(pc, provider, tagePred);

        ref LoopEntry le = ref _loop[LoopIdx(pc)];
        if (le.Tag == (ushort)LoopTag(pc) && le.Confident)
            pred = le.CurrentIter < le.LearnedIter; // taken while not yet at trip count

        ulong target = pred
            ? _btb.TryGetValue(pc, out ulong t) ? t : pc + 4
            : pc + 4;
        return new BranchPrediction(pred, target);
    }

    /// <inheritdoc />
    public virtual void Update(ulong pc, bool taken, ulong actualTarget) {
        if (taken) _btb[pc] = actualTarget;

        // Train against committed (predict-time) history: swap it into Ghr for the table
        // lookups, then restore the working history afterward. In non-speculative mode the
        // two are equal, so this is a no-op swap, and behavior is unchanged.
        ulong working = Ghr;
        Ghr = _committedGhr;

        TageLookup(pc, out int provider, out bool provPred, out bool altPred);
        int preScore = TageScore(pc, provider);
        bool preLoopConfident = _loop[LoopIdx(pc)].Tag == (ushort)LoopTag(pc) && _loop[LoopIdx(pc)].Confident;
        if (!SuppressTageUpdate(pc)) {
            UpdateTage(pc, taken, provider, provPred, altPred);
            UpdateLoop(pc, taken);
        }

        OnAfterUpdate(pc, taken, provPred, preScore, preLoopConfident);

        _committedGhr = ((_committedGhr << 1) | (taken ? 1UL : 0UL)) & ((1UL << LTageBp.MaxHist) - 1);
        // Non-speculative: keep the working history in lock-step with committed (old
        // behavior). Speculative: leave it as fetch advanced it; flush restores it.
        Ghr = _speculative ? working : _committedGhr;
    }

    /// <inheritdoc />
    public virtual void SpeculativeHistoryUpdate(ulong pc, bool predictedTaken) {
        _speculative = true;
        Ghr = ((Ghr << 1) | (predictedTaken ? 1UL : 0UL)) & ((1UL << LTageBp.MaxHist) - 1);
    }

    /// <inheritdoc />
    public virtual void RecoverSpeculativeHistory() => Ghr = _committedGhr;

    /// <inheritdoc />
    // The whole TAGE family indexes off Ghr (folded histories are recomputed on the fly from it),
    // so a single ulong checkpoint of Ghr is a complete history snapshot. No local component.
    public virtual BranchHistoryCheckpoint CaptureHistory(ulong pc) => new(Ghr);

    /// <inheritdoc />
    public virtual void RestoreHistory(in BranchHistoryCheckpoint checkpoint, ulong pc, bool actualTaken) {
        _speculative = true;
        Ghr = ((checkpoint.Global << 1) | (actualTaken ? 1UL : 0UL)) & ((1UL << LTageBp.MaxHist) - 1);
    }

    // ── Extension points for subclasses ──────────────────────────────────────

    /// <summary>
    ///     Returns the score of the prediction for the given PC.
    /// </summary>
    /// <param name="pc">Program counter.</param>
    /// <param name="provider">
    ///     Index of the provider table, or -1 if the PC is in the base table.
    /// </param>
    /// <returns></returns>
    protected int TageScore(ulong pc, int provider) {
        if (provider >= 0) {
            byte c = _tage[provider][TageIdx(pc, provider)].Ctr;
            return c * 2 - 7; // 0..7 → -7..+7
        }

        byte b = _base[BaseIdx(pc)];
        return b * 2 - 3; // 0..3 → -3..+3
    }

    /// <summary>
    ///     Returns the prediction for the given PC.
    /// </summary>
    /// <param name="pc">
    ///     Program counter.
    /// </param>
    /// <param name="provider">
    ///     Index of the provider table, or -1 if the PC is in the base table.
    /// </param>
    /// <param name="tagePred">
    ///     TAGE prediction, or false if the PC is in the base table.
    /// </param>
    /// <returns></returns>
    protected virtual bool ResolvePrediction(ulong pc, int provider, bool tagePred) => tagePred;

    /// <summary>
    ///     Called after a branch update.
    /// </summary>
    /// <param name="pc">Program counter.</param>
    /// <param name="taken">Branch was taken.</param>
    /// <param name="provPred">
    ///     TAGE prediction, or false if the PC is in the base table.
    /// </param>
    /// <param name="preScore">
    ///     Score of the prediction before the update.
    /// </param>
    /// <param name="loopWasConfident">
    ///     True if the loop predictor was confident before the update.
    /// </param>
    protected virtual void OnAfterUpdate(
        ulong pc,
        bool taken,
        bool provPred,
        int preScore,
        bool loopWasConfident
    ) { }

    /// <summary>
    ///     When true, skips the TAGE table and loop table writes (and, transitively, any
    ///     subclass's <see cref="OnAfterUpdate" /> training that itself checks this hook — see
    ///     <see cref="TageScLBp.OnAfterUpdate" />) for this branch, while global history
    ///     still advances normally. Used by <see cref="BullseyeBp" /> to stop polluting
    ///     the TAGE-SC-L substrate for branches its perceptron layer has taken over, without
    ///     breaking history-based indexing for every other branch. Default false.
    /// </summary>
    protected virtual bool SuppressTageUpdate(ulong pc) => false;

    /// <summary>
    ///     Usefulness counter of the given PC's provider-table entry, or 0 if the PC resolved to
    ///     the untagged base predictor. Exposed for subclasses (e.g. <see cref="BullseyeBp" />'s
    ///     TAGE confidence gate) that need the raw usefulness bit, not just <see cref="TageScore" />.
    /// </summary>
    protected byte TageUsefulness(ulong pc, int provider) =>
        provider >= 0 ? _tage[provider][TageIdx(pc, provider)].U : (byte)0;

    // ── TAGE internals ────────────────────────────────────────────────────────

    // Scan tables shortest→longest to find provider (longest match) and alt provider.
    private void TageLookup(
        ulong pc,
        out int provider,
        out bool provPred,
        out bool altPred
    ) {
        bool basePred = _base[BaseIdx(pc)] >= 2;
        provider = -1;
        provPred = basePred;
        altPred = basePred;

        for (var t = 0; t < LTageBp.NumTables; t++) {
            ref TageEntry e = ref _tage[t][TageIdx(pc, t)];
            if (!e.Valid || e.Tag != (ushort)TageTag(pc, t)) continue;
            altPred = provPred;
            provider = t;
            provPred = e.Ctr >= 4;
        }
    }

    private void UpdateTage(
        ulong pc,
        bool taken,
        int provider,
        bool provPred,
        bool altPred
    ) {
        // Update provider counter and usefulness.
        if (provider >= 0) {
            ref TageEntry e = ref _tage[provider][TageIdx(pc, provider)];
            Sat3(ref e.Ctr, taken);
            if (provPred != altPred) {
                if (provPred == taken) {
                    if (e.U < 3) e.U++;
                }
                else {
                    if (e.U > 0) e.U--;
                }
            }
        }
        else { Sat2(ref _base[BaseIdx(pc)], taken); }

        // On misprediction, allocate in shortest eligible longer-history table.
        if (provPred == taken) return;

        var alloc = false;
        for (int t = provider + 1; t < LTageBp.NumTables; t++) {
            ref TageEntry e = ref _tage[t][TageIdx(pc, t)];
            if (e.U == 0) {
                e.Tag = (ushort)TageTag(pc, t);
                e.Ctr = taken ? (byte)4 : (byte)3; // weakly in correct direction
                e.U = 0;
                e.Valid = true;
                alloc = true;
                break;
            }
        }

        // If no slot was free, decay usefulness to free one in a future miss.
        if (!alloc)
            for (int t = provider + 1; t < LTageBp.NumTables; t++) {
                ref TageEntry e = ref _tage[t][TageIdx(pc, t)];
                if (e.U > 0) e.U--;
            }
    }

    // ── Loop predictor internals ──────────────────────────────────────────────

    private void UpdateLoop(ulong pc, bool taken) {
        int li = LoopIdx(pc);
        var lt = (ushort)LoopTag(pc);
        ref LoopEntry le = ref _loop[li];

        if (le.Tag == lt) {
            if (taken) { le.CurrentIter++; }
            else {
                if (!le.Confident) {
                    if (le.LearnedIter == 0 || le.LearnedIter == le.CurrentIter) {
                        le.LearnedIter = le.CurrentIter;
                        if (le.ConfCount < 255) le.ConfCount++;
                        if (le.ConfCount >= LTageBp.LoopConfidence) le.Confident = true;
                    }
                    else {
                        // Inconsistent trip count — restart tracking at this slot.
                        le = new LoopEntry { Tag = lt, };
                    }
                }
                else if (le.CurrentIter != le.LearnedIter) {
                    // Trip count changed — lose confidence and re-learn.
                    le.Confident = false;
                    le.ConfCount = 0;
                    le.LearnedIter = le.CurrentIter;
                }

                le.CurrentIter = 0;
            }
        }
        else if (taken) {
            // First time seeing a taken branch at this slot: start tracking.
            le = new LoopEntry { Tag = lt, CurrentIter = 1, };
        }
    }

    // ── Index / tag helpers ───────────────────────────────────────────────────

    private static int BaseIdx(ulong pc) =>
        (int)((pc >> 2) & ((1u << LTageBp.BaseIndexBits) - 1));

    /// <summary>Returns the bimodal (T0) prediction for <paramref name="pc" />.</summary>
    protected bool BimodalPrediction(ulong pc) => _base[BaseIdx(pc)] >= 2;

    private int TageIdx(ulong pc, int t) {
        int folded = FoldHist(LTageBp.HistLengths[t], LTageBp.TableIndexBits);
        return ((int)(pc >> 2) ^ folded) & ((1 << LTageBp.TableIndexBits) - 1);
    }

    // Two independent folds of different history lengths produce two tag halves,
    // reducing aliasing between branches that share the same PC index.
    /// <summary>Computes the partial TAGE tag for branch <paramref name="pc" /> in table <paramref name="t" />.</summary>
    protected int TageTag(ulong pc, int t) {
        int f1 = FoldHist(LTageBp.HistLengths[t], LTageBp.TagWidth);
        int f2 = FoldHist(LTageBp.HistLengths[t] - 1, LTageBp.TagWidth - 1);
        return ((int)(pc >> 2) ^ f1 ^ (f2 << 1)) & ((1 << LTageBp.TagWidth) - 1);
    }

    private static int LoopIdx(ulong pc) =>
        (int)((pc >> 2) & ((1u << LTageBp.LoopIndexBits) - 1));

    // Mix high PC bits into low bits to separate tag from index.
    private static int LoopTag(ulong pc) {
        ulong x = pc >> 2;
        return (int)((x ^ (x >> LTageBp.LoopIndexBits)) & ((1u << LTageBp.LoopTagWidth) - 1));
    }

    // XOR-fold the lowest `histLen` bits of _ghr into `outBits` bits.
    // Masking _ghr to histLen bits first ensures no garbage above.
    private int FoldHist(int histLen, int outBits) {
        ulong hist = Ghr & ((1UL << histLen) - 1);
        int mask = (1 << outBits) - 1;
        var res = 0;
        for (var sh = 0; sh < histLen; sh += outBits) res ^= (int)((hist >> sh) & (ulong)mask);
        return res;
    }

    // ── Saturating counter helpers ────────────────────────────────────────────

    private static void Sat2(ref byte c, bool taken) {
        switch (taken) {
            case true when c < 3:  c++; break;
            case false when c > 0: c--; break;
        }
    }

    private static void Sat3(ref byte c, bool taken) {
        switch (taken) {
            case true when c < 7:  c++; break;
            case false when c > 0: c--; break;
        }
    }

    // ── Entry structs ─────────────────────────────────────────────────────────

    private struct TageEntry {
        public ushort Tag;
        public byte Ctr; // 3-bit prediction counter (taken ≥ 4)
        public byte U;   // 2-bit usefulness counter
        public bool Valid;
    }

    private struct LoopEntry {
        public ushort Tag;
        public ushort LearnedIter; // stable trip count
        public ushort CurrentIter; // iterations since last exit
        public byte ConfCount;     // consistent exits seen
        public bool Confident;
    }
}